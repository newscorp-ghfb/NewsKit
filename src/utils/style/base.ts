import {Theme, BreakpointKeys} from '../../theme';
import {isResponsive, getMediaQueryFromTheme} from '../responsive-helpers';
import {filterObject} from '../filter-object';
import {getToken} from '../get-token';
import {ThemeProp} from '../style-types';
import {CSSQuery, MQ, ResponsiveValue} from './types';
import {isNonThemeValueAllowed, isValidUnit} from './utils';
import {CSSObject} from './emotion';

export type FromThemeCallback = (cssValue: string) => CSSObject | string;

export const getDefaultedValue = <
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  FromThemeUtil extends (...args: any) => any
>(
  getPresetFromThemeUtil: FromThemeUtil,
  presetType: string,
  cssProp?: string | FromThemeCallback,
) => <Props extends ThemeProp>(
  defaultPath: string | undefined,
  overridePath: string | false = '',
  option?: Parameters<FromThemeUtil>[2],
) => (props: Props) =>
  getPresetFromThemeUtil(
    getToken(props, defaultPath, overridePath, presetType),
    undefined,
    cssProp || option,
  )(props);

export const getValueFromTheme = <ThemeToken extends string>(
  themeKey: keyof Theme,
) => <Props extends ThemeProp>(
  defaultToken?: MQ<ThemeToken>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => (props: Props) => {
  const canHaveNonThemeValue = isNonThemeValueAllowed(themeKey);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const section = props.theme[themeKey] as any;
  const token = (customProp && props[customProp]) || defaultToken;

  if (section && section[token]) {
    return section[token];
  }
  if (canHaveNonThemeValue && isValidUnit(themeKey, token)) {
    return token;
  }

  return '';
};

export const getResponsiveValueFromTheme = <ThemeToken extends string>(
  themeKey: keyof Theme,
) => <Props extends ThemeProp>(
  defaultToken?: ResponsiveValue<ThemeToken>,
  customProp?: Exclude<keyof Props, 'theme'>,
) => ({theme, ...props}: Props) => {
  const section = theme[themeKey] as Record<ThemeToken, unknown>;
  const propKeys = (customProp && props[customProp]) || defaultToken;
  const {breakpoints} = theme;
  const canHaveNonThemeValue = isNonThemeValueAllowed(themeKey);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const isMQTokenArray = (v: any): string[] =>
    v.length > 0 &&
    v.every((token: ThemeToken) => section[token as ThemeToken]);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const mapTokensArray = (v: any): string[] =>
    v.map((token: ThemeToken) => section[token as ThemeToken]).join(' ');
  if (isResponsive(propKeys, breakpoints)) {
    // We have a breakpoints object...
    // Convert breakpoints to array and order them
    const breakpointKeys = Object.keys(breakpoints) as BreakpointKeys[];
    const mq = breakpointKeys.sort((a, b) => breakpoints[a] - breakpoints[b]);
    const presetKeys: [BreakpointKeys, ThemeToken][] = Object.entries(
      filterObject(propKeys, mq),
    ).sort(
      ([a], [b]) =>
        mq.indexOf(a as BreakpointKeys) - mq.indexOf(b as BreakpointKeys),
    ) as any; // eslint-disable-line @typescript-eslint/no-explicit-any
    const cssMediaQueryObject = presetKeys
      .filter(
        // Exclude invalid breakpoints and theme section keys
        ([breakpointKey, presetKey]) =>
          presetKey &&
          (section[presetKey] ||
            (themeKey === 'spacePresets' &&
              isMQTokenArray(presetKey.split(' '))) ||
            (canHaveNonThemeValue && isValidUnit(themeKey, presetKey))) &&
          breakpointKeys.includes(breakpointKey),
      )
      .reduce((acc, [key, presetKey], index, arr) => {
        /* istanbul ignore next */
        let preset = '' as Record<ThemeToken, unknown>[ThemeToken];
        const MQtokens =
          typeof presetKey === 'string' && (presetKey as string).split(' ');
        if (themeKey === 'spacePresets' && isMQTokenArray(MQtokens)) {
          preset = mapTokensArray(MQtokens);
        } else {
          preset =
            section[presetKey] ||
            (canHaveNonThemeValue &&
              isValidUnit(themeKey, presetKey) &&
              presetKey);
        }
        // Get next key to set the max. This stops styles overlapping when they
        // shouldn't by explicitly setting them for the range they need to apply on.
        const nextKey = arr[index + 1] ? arr[index + 1][0] : undefined;
        const mediaQuery = getMediaQueryFromTheme(
          key,
          nextKey,
        )({
          theme,
        });
        acc[mediaQuery] = preset;
        return acc;
      }, {} as Record<string, unknown>);

    const containerKeys = (propKeys.rules || []) as CSSQuery<ThemeToken>[];

    const cssContainerQueryObject =
      containerKeys &&
      containerKeys.reduce((acc, query) => {
        const {rule, value} = query;
        let preset = '' as Record<ThemeToken, unknown>[ThemeToken];
        const MQtokens =
          typeof value === 'string' && (value as string).split(' ');
        if (themeKey === 'spacePresets' && isMQTokenArray(MQtokens)) {
          preset = mapTokensArray(MQtokens);
        } else {
          preset =
            section[value] ||
            (canHaveNonThemeValue && isValidUnit(themeKey, value) && value);
        }
        acc[rule] = preset;
        return acc;
      }, {} as Record<string, unknown>);

    return Object.entries({...cssMediaQueryObject, ...cssContainerQueryObject});
  }

  const noMQtokens =
    typeof propKeys === 'string' && themeKey === 'spacePresets'
      ? (propKeys as string).split(' ')
      : [];
  if (isMQTokenArray(noMQtokens)) {
    return mapTokensArray(noMQtokens);
  }

  if (propKeys && section[propKeys as ThemeToken]) {
    return section[propKeys as ThemeToken];
  }
  if (canHaveNonThemeValue && propKeys && isValidUnit(themeKey, propKeys)) {
    return propKeys;
  }
  return '';
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const isNotNullOrEmpty = (v: any): boolean =>
  v !== null && v !== undefined && v !== '';

export const getXFromTheme = (themeKey: keyof Theme) => <
  Props extends ThemeProp
>(
  cssProperty: string | FromThemeCallback,
  defaultToken: ResponsiveValue<string>,
) => (props: Props) => {
  const value = getResponsiveValueFromTheme(themeKey)(defaultToken)(
    props,
  ) as Array<[string, string]>;

  if (Array.isArray(value)) {
    return value.reduce((acc, [mq, preset]) => {
      let mqValue;
      if (typeof cssProperty === 'string') {
        mqValue = {[cssProperty]: preset};
      }
      if (typeof cssProperty === 'function') {
        mqValue = cssProperty(preset);
      }
      /* istanbul ignore next */
      if (mqValue) {
        acc[mq] = mqValue;
      }
      return acc;
    }, {} as CSSObject);
  }

  if (typeof cssProperty === 'string' && isNotNullOrEmpty(value)) {
    return {[cssProperty]: value};
  }

  if (typeof cssProperty === 'function' && isNotNullOrEmpty(value)) {
    return cssProperty(value);
  }

  return value;
};

export const getResponsiveX = (themeKey: keyof Theme) => (
  cssProperty: string | FromThemeCallback,
  defaultPath: string,
  overridePath: string,
  defaultsObjectKey: string,
) => <Props extends ThemeProp>(props: Props) => {
  const token = getToken(props, defaultPath, overridePath, defaultsObjectKey);
  return getXFromTheme(themeKey)(cssProperty, token)(props);
};
