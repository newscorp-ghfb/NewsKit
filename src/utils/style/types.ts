export type MQPartial<T> = Partial<{
  xs: T;
  sm: T;
  md: T;
  lg: T;
  xl: T;
}>;

export type MQ<T> = T | MQPartial<T>;

/*
Allows you define values at set breakpoints, or define custom media & container queries with the "rules" object
*/
export type ResponsiveValue<T> = MQ<T> | CSSQueryRules<T>;

/*
  @param rule - The media or container query definition
  @param value - CSS value or token
*/
export type CSSQuery<T> = {
  rule: `@media ${string}` | `@container ${string}`;
  value: T;
};

export type CSSQueryRules<T> = {
  rules?: CSSQuery<T>[];
};

export type ContainerQueryProps = {
  containerType?:
    | `normal`
    | `inline-size`
    | `size`
    | `inherit`
    | `initial`
    | `unset`
    | `revert`
    | `revert-layer`;
  containerName?: string;
};
