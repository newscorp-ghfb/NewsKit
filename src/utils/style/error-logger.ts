// eslint-disable-next-line no-console
export const errorLogger = console.error.bind(console);
