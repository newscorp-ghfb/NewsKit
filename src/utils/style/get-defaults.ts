import {getDefaultedValue, getResponsiveX} from './base';
import {
  getBorderFromTheme,
  getSizingFromTheme,
  getSpacingFromTheme,
  getSpacingInsetFromTheme,
  getTypographyPresetFromTheme,
} from './getters';

export const getTypographyPreset = getDefaultedValue(
  getTypographyPresetFromTheme,
  'typographyPreset',
);

/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSpacingInlineHorizontal instead
 */
export const getSpacingInlineHorizontal = getDefaultedValue(
  getSpacingFromTheme,
  'spaceInline',
  'marginRight',
);

/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSpacingInlineVertical instead
 */
export const getSpacingInlineVertical = getDefaultedValue(
  getSpacingFromTheme,
  'spaceInline',
  'marginBottom',
);

/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSpacingStackHorizontal instead
 */
export const getSpacingStackHorizontal = getDefaultedValue(
  getSpacingFromTheme,
  'spaceStack',
  'marginBottom',
);

/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSpacingStackVertical instead
 */
export const getSpacingStackVertical = getDefaultedValue(
  getSpacingFromTheme,
  'spaceStack',
  'marginRight',
);

/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSpace instead
 */
export const getSpacingInset = getDefaultedValue(
  getSpacingInsetFromTheme,
  'spaceInset',
);

/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSize instead
 */
export const getSizing = getDefaultedValue(getSizingFromTheme, 'sizing');
/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSize instead
 */
export const getMinHeight = getDefaultedValue(getSizingFromTheme, 'minHeight');
/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSize instead
 */
export const getMinWidth = getDefaultedValue(getSizingFromTheme, 'minWidth');
/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSize instead
 */
export const getHeight = getDefaultedValue(getSizingFromTheme, 'height');
/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSize instead
 */
export const getWidth = getDefaultedValue(getSizingFromTheme, 'width');
/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSize instead
 */
export const getSize = getDefaultedValue(getSizingFromTheme, 'size');
/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSize instead
 */
export const getMaxSize = getDefaultedValue(getSizingFromTheme, 'maxSize');
/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSize instead
 */
export const getMinSize = getDefaultedValue(getSizingFromTheme, 'minSize');
/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveSpace instead
 */
export const getSpace = getDefaultedValue(getSpacingFromTheme, 'space');
/**
 * @deprecated This method has been deprecated and will be removed in a future release, use getResponsiveBorder instead
 */
export const getWeight = getDefaultedValue(getBorderFromTheme, 'weight');

export const getResponsiveSpace = getResponsiveX('spacePresets');
export const getResponsiveSize = getResponsiveX('sizing');
export const getResponsiveMotion = getResponsiveX('motions');
export const getResponsiveBorder = getResponsiveX('borders');

const createResponsiveSpace = (
  cssAttribute: string,
  defaultsObjectKey: string,
) => (defaultPath: string, overridePath?: string) =>
  getResponsiveSpace(
    cssAttribute,
    defaultPath,
    overridePath || '',
    defaultsObjectKey,
  );

export const getResponsiveSpacingInlineHorizontal = createResponsiveSpace(
  'marginRight',
  'spaceInline',
);
export const getResponsiveSpacingInlineVertical = createResponsiveSpace(
  'marginBottom',
  'spaceInline',
);
export const getResponsiveSpacingStackHorizontal = createResponsiveSpace(
  'marginBottom',
  'spaceStack',
);
export const getResponsiveSpacingStackVertical = createResponsiveSpace(
  'marginRight',
  'spaceStack',
);
export const getResponsiveSpacingInset = createResponsiveSpace(
  'padding',
  'spaceInset',
);
