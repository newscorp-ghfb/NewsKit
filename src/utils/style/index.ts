export * from './base';
export * from './emotion';
export * from './getters';
export * from './get-defaults';
export * from './style-preset';
export * from './transition-preset';
export * from './types';
