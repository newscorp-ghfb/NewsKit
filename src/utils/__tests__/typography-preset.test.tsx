import {renderToFragmentWithTheme} from '../../test/test-utils';
import {createTheme, Theme} from '../../theme';
import {getTypographyPresetFromTheme, styled, MQ} from '../style';

interface TestTextProp {
  typographyPreset: MQ<string>;
  options?: {
    withCrop: boolean;
  };
}

const myCustomTheme = createTheme({
  name: 'my-custom-typography',
  overrides: {
    fonts: {
      fontFamily100: {
        fontFamily: '"Custom-FontName", sans-serif',
        fontMetrics: {
          fontWeight010: {
            capHeight: 697,
            ascent: 1050,
            descent: -350,
            lineGap: 100,
            unitsPerEm: 1000,
          },
        },
      },
      fontFamily200: {
        fontFamily: '"Custom-FontName2", sans-serif',
        cropConfig: {
          top: 9,
          bottom: 9,
        },
      },
    },
    typographyPresets: {
      editorialParagraph030: {
        fontSmooth: 'always',
        fontStretch: 'condensed',
      },
      editorialParagraph100: {
        fontFamily: 'Non-existent-font-family, serif',
        fontSize: '{{fonts.fontSize040}}',
        lineHeight: '{{fonts.fontLineHeight040}}',
        fontWeight: '{{fonts.fontWeight010}}',
        letterSpacing: '{{fonts.fontLetterSpacing010}}',
      },
      editorialParagraph300: {
        fontFamily: '{{fonts.fontFamily100.fontFamily}}',
        fontSize: '{{fonts.fontSize040}}',
        lineHeight: '{{fonts.fontLineHeight040}}',
        fontWeight: '{{fonts.fontWeight010}}',
        letterSpacing: '{{fonts.fontLetterSpacing010}}',
      },
      editorialParagraph400: {
        fontFamily: '{{fonts.fontFamily200.fontFamily}}',
        fontSize: '{{fonts.fontSize040}}',
        lineHeight: '{{fonts.fontLineHeight040}}',
        fontWeight: '{{fonts.fontWeight010}}',
        letterSpacing: '{{fonts.fontLetterSpacing010}}',
      },
    },
  },
});

const TestText = styled.p<TestTextProp>`
  ${getTypographyPresetFromTheme('editorialParagraph030', 'typographyPreset')}
`;

describe('TypographyPreset', () => {
  test('renders default', () => {
    const fragment = renderToFragmentWithTheme(TestText);
    expect(fragment).toMatchSnapshot();
  });

  test('renders with override editorialParagraph010', () => {
    const fragment = renderToFragmentWithTheme(TestText, {
      typographyPreset: 'editorialParagraph010',
    });
    expect(fragment).toMatchSnapshot();
  });

  test('renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, and editorialParagraph030 for md breakpoints', () => {
    const fragment = renderToFragmentWithTheme(TestText, {
      typographyPreset: {
        xs: 'editorialParagraph010',
        sm: 'editorialParagraph020',
        md: 'editorialParagraph030',
      },
    });
    expect(fragment).toMatchSnapshot();
  });

  test('renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, and wrong for md breakpoints', () => {
    const fragment = renderToFragmentWithTheme(TestText, {
      typographyPreset: {
        xs: 'editorialParagraph010',
        sm: 'editorialParagraph020',
        md: 'wrong',
      } as any,
    });
    expect(fragment).toMatchSnapshot();
  });

  test('renders responsive override editorialParagraph010 for xs and editorialParagraph030 for md breakpoints', () => {
    const fragment = renderToFragmentWithTheme(TestText, {
      typographyPreset: {
        xs: 'editorialParagraph010',
        md: 'editorialParagraph030',
      },
    });
    expect(fragment).toMatchSnapshot();
  });

  test('renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, editorialParagraph030 for md, editorialHeading010 for lg and xl breakpoints and ignore wrong prop', () => {
    const fragment = renderToFragmentWithTheme(TestText, {
      typographyPreset: {
        xs: 'editorialParagraph010',
        sm: 'editorialParagraph020',
        md: 'editorialParagraph030',
        lg: 'editorialHeading010',
        xl: 'editorialHeading030',
        wrong: 'editorialHeading020',
      } as any,
    });
    expect(fragment).toMatchSnapshot();
  });

  test('renders with fontSmooth and fontStretch', () => {
    const fragment = renderToFragmentWithTheme(
      TestText,
      undefined,
      myCustomTheme,
    );
    expect(fragment).toMatchSnapshot();
  });

  test('when withCrop is true renders cropped text', () => {
    const TestText1 = styled.p<TestTextProp>`
      ${getTypographyPresetFromTheme(
        'editorialParagraph030',
        'typographyPreset',
        {withCrop: true},
      )}
    `;

    const fragment = renderToFragmentWithTheme(TestText1);
    expect(fragment).toMatchSnapshot();
  });

  test('when withCrop is true but such font does not exist renders standard text', () => {
    const TestText1 = styled.p<TestTextProp>`
      ${getTypographyPresetFromTheme(
        'editorialParagraph100',
        'typographyPreset',
        {withCrop: true},
      )}
    `;

    const fragment = renderToFragmentWithTheme(
      TestText1,
      undefined,
      myCustomTheme,
    );
    expect(fragment).toMatchSnapshot();
  });

  test('when withCrop is true & cropping functions work with fontMetrics', () => {
    const TestText1 = styled.p<TestTextProp>`
      ${getTypographyPresetFromTheme(
        'editorialParagraph300',
        'typographyPreset',
        {withCrop: true},
      )}
    `;

    const fragment = renderToFragmentWithTheme(
      TestText1,
      undefined,
      myCustomTheme,
    );
    expect(fragment).toMatchSnapshot();
  });

  test('when withCrop is true cropping functions work with cropConfig', () => {
    const TestText1 = styled.p<TestTextProp>`
      ${getTypographyPresetFromTheme(
        'editorialParagraph400',
        'typographyPreset',
        {withCrop: true},
      )}
    `;

    const fragment = renderToFragmentWithTheme(
      TestText1,
      undefined,
      myCustomTheme,
    );
    expect(fragment).toMatchSnapshot();
  });

  test('returns typographyPreset with text crop CSS when fontMetrics exist for fontFamily/fontWeight', () => {
    const typographyPreset = getTypographyPresetFromTheme(
      'editorialParagraph030',
      undefined,
      {withCrop: true},
    )({
      theme: ({
        fonts: {
          fontWeight020: 500,
          fontFamily010: {
            fontFamily: 'Font Family Name',
            fontMetrics: {
              fontWeight020: {
                capHeight: 697,
                ascent: 1050,
                descent: -350,
                lineGap: 100,
                unitsPerEm: 1000,
              },
            },
          },
        },
        typographyPresets: {
          editorialParagraph030: {
            fontFamily: 'Font Family Name',
            fontWeight: 500,
            fontSize: '12px',
            lineHeight: 1,
          },
        },
      } as unknown) as Theme,
    });
    expect(typographyPreset).toEqual({
      fontFamily: 'Font Family Name',
      fontWeight: 500,
      fontSize: '12px',
      lineHeight: '12px',
      '::before': {
        content: "''",
        marginBottom: '-0.153em',
        display: 'block',
      },
      '::after': {
        content: "''",
        marginTop: '-0.15em',
        display: 'block',
      },
      padding: '0.5px 0px',
    });
  });

  test('returns typographyPreset with default text crop CSS when fontMetrics do not exist for fontFamily/fontWeight', () => {
    const typographyPreset = getTypographyPresetFromTheme(
      'editorialParagraph030',
      undefined,
      {withCrop: true},
    )({
      theme: ({
        fonts: {
          fontWeight020: 500,
          fontFamily010: {
            fontFamily: 'Font Family Name',
            fontMetrics: {
              // not defined for fontWeight020
              fontWeight010: {
                capHeight: 692,
                ascent: 935,
                descent: -265,
                lineGap: 0,
                unitsPerEm: 1000,
              },
            },
          },
        },
        typographyPresets: {
          editorialParagraph030: {
            fontFamily: 'Font Family Name',
            fontWeight: 500,
            fontSize: '12px',
            lineHeight: 1,
          },
        },
      } as unknown) as Theme,
    });
    expect(typographyPreset).toEqual({
      fontFamily: 'Font Family Name',
      fontWeight: 500,
      fontSize: '12px',
      lineHeight: '12px',
      '::before': {
        content: "''",
        marginBottom: '-0.143em',
        display: 'block',
      },
      '::after': {
        content: "''",
        marginTop: '-0.165em',
        display: 'block',
      },
      padding: '0.5px 0px',
    });
  });

  test('returns typographyPreset without text crop CSS when fontMetrics do not exist for fontFamily/fontWeight or for default fontWeight', () => {
    jest.spyOn(console, 'warn').mockImplementation();
    const typographyPreset = getTypographyPresetFromTheme(
      'editorialParagraph030',
      undefined,
      {withCrop: true},
    )({
      theme: ({
        fonts: {
          fontWeight020: 500,
          fontFamily010: {
            fontFamily: 'Font Family Name',
            fontMetrics: {
              // default fontWeight010 not defined
            },
          },
        },
        typographyPresets: {
          editorialParagraph030: {
            fontFamily: 'Font Family Name',
            fontWeight: 500,
            fontSize: 12,
          },
        },
      } as unknown) as Theme,
    });
    expect(typographyPreset).toEqual({
      fontFamily: 'Font Family Name',
      fontWeight: 500,
      fontSize: 12,
    });
    // eslint-disable-next-line no-console
    expect(console.warn).toHaveBeenCalledWith(
      "No default fontMetrics found for 'Font Family Name'.",
    );
  });
});
