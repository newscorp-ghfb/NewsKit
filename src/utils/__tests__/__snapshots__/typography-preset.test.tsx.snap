// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypographyPreset renders default 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: "DM Sans",sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0;
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders responsive override editorialParagraph010 for xs and editorialParagraph030 for md breakpoints 1`] = `
<DocumentFragment>
  @media screen and (max-width: 767px) {
  .emotion-0 {
    font-family: "DM Sans",sans-serif;
    font-size: 14px;
    line-height: 1.5;
    font-weight: 400;
    letter-spacing: 0;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0 {
    font-family: "DM Sans",sans-serif;
    font-size: 18px;
    line-height: 1.5;
    font-weight: 400;
    letter-spacing: 0;
  }
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, and editorialParagraph030 for md breakpoints 1`] = `
<DocumentFragment>
  @media screen and (max-width: 479px) {
  .emotion-0 {
    font-family: "DM Sans",sans-serif;
    font-size: 14px;
    line-height: 1.5;
    font-weight: 400;
    letter-spacing: 0;
  }
}

@media screen and (min-width: 480px) and (max-width: 767px) {
  .emotion-0 {
    font-family: "DM Sans",sans-serif;
    font-size: 16px;
    line-height: 1.5;
    font-weight: 400;
    letter-spacing: 0;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0 {
    font-family: "DM Sans",sans-serif;
    font-size: 18px;
    line-height: 1.5;
    font-weight: 400;
    letter-spacing: 0;
  }
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, and wrong for md breakpoints 1`] = `
<DocumentFragment>
  @media screen and (max-width: 479px) {
  .emotion-0 {
    font-family: "DM Sans",sans-serif;
    font-size: 14px;
    line-height: 1.5;
    font-weight: 400;
    letter-spacing: 0;
  }
}

@media screen and (min-width: 480px) {
  .emotion-0 {
    font-family: "DM Sans",sans-serif;
    font-size: 16px;
    line-height: 1.5;
    font-weight: 400;
    letter-spacing: 0;
  }
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders responsive override editorialParagraph010 for xs, editorialParagraph020 for sm, editorialParagraph030 for md, editorialHeading010 for lg and xl breakpoints and ignore wrong prop 1`] = `
<DocumentFragment>
  @media screen and (max-width: 479px) {
  .emotion-0 {
    font-family: "DM Sans",sans-serif;
    font-size: 14px;
    line-height: 1.5;
    font-weight: 400;
    letter-spacing: 0;
  }
}

@media screen and (min-width: 480px) and (max-width: 767px) {
  .emotion-0 {
    font-family: "DM Sans",sans-serif;
    font-size: 16px;
    line-height: 1.5;
    font-weight: 400;
    letter-spacing: 0;
  }
}

@media screen and (min-width: 768px) {
  .emotion-0 {
    font-family: "DM Sans",sans-serif;
    font-size: 18px;
    line-height: 1.5;
    font-weight: 400;
    letter-spacing: 0;
  }
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders with fontSmooth and fontStretch 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: "DM Sans",sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0;
  font-smooth: always;
  font-stretch: condensed;
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset renders with override editorialParagraph010 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: "DM Sans",sans-serif;
  font-size: 14px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0;
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset when withCrop is true & cropping functions work with fontMetrics 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: "Custom-FontName",sans-serif;
  font-size: 18px;
  line-height: 27px;
  font-weight: 400;
  letter-spacing: -0.5px;
  padding: 0.5px 0px;
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.403em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.4em;
  display: block;
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset when withCrop is true but such font does not exist renders standard text 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: Non-existent-font-family,serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: -0.5px;
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset when withCrop is true cropping functions work with cropConfig 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: "Custom-FontName2",sans-serif;
  font-size: 18px;
  line-height: 1.5555555555555556;
  font-weight: 400;
  letter-spacing: -0.5px;
  padding: 1px 0px;
}

.emotion-0::before {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-bottom: -0.4590277777777778em;
}

.emotion-0::after {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-top: -0.4590277777777778em;
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`TypographyPreset when withCrop is true renders cropped text 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: "DM Sans",sans-serif;
  font-size: 18px;
  line-height: 27px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.391em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.409em;
  display: block;
}

<p
    class="emotion-0"
  />
</DocumentFragment>
`;
